### PR TITLE
Kuc unit test attachment

### PR DESCRIPTION
**Pull request by phongnhat19**

### CODE REVIEW

File: src/js/Attachment/__test__/showError.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  test('showError is called successfully', () => {
+    const ERROR_MESSAGE_TEXT = 'Error.';
+    const myAttachment = new Attachment({isErrorVisible: false});
+    const container = myAttachment.render();
+    myAttachment.setErrorMessage(ERROR_MESSAGE_TEXT);
+    myAttachment.showError();
+    const errorDOM = container.getElementsByClassName('kuc-attachment-file-error');
+    expect(errorDOM.length).toEqual(1);
+    expect(errorDOM[0]).toBeInstanceOf(HTMLDivElement);
+    expect((errorDOM[0] as HTMLDivElement).style.display).toEqual('block');
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/showError.test.ts]: Should change to simple code
expect((errorDOM[0] as HTMLDivElement).style.display).toEqual('block'); 
==> expect(errorDOM[0]).toHaveStyle('display: block');

File: src/js/Attachment/__test__/setFiles.test.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    myAttachment.setFiles([{name: FILE_NAME, size: FILE_SIZE}]);
+
+    // Verify files prop
+    const files = myAttachment.getFiles();
+    expect(files).toBeInstanceOf(Array);
+    expect(files!.length).toEqual(1);
+    expect(files![0]).toHaveProperty('name');
+    expect(files![0]).toHaveProperty('size');
+    expect(files![0].name).toEqual(FILE_NAME);
+    expect(files![0].size).toEqual(FILE_SIZE);
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/setFiles.test.ts]: Line no need:
expect(files![0]).toHaveProperty('name');
expect(files![0]).toHaveProperty('size');

File: src/js/Attachment/__test__/onDrop.test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    });
+
+    const newFile = new File([''], 'file2.png', {
+      type: 'image/png',
+    });
+
+    const dropHandler = (files: any[]) => {
+      expect(files.length).toEqual(2);
+      expect(files[1]).toBe(newFile);
+    };
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/onDrop.test.ts]: Seem this test is not effected.

Change expect(files.length).toEqual(1); --> expect(files.length).toEqual(344);
Result: test suite is passed
Expect: error is shown

File: src/js/Attachment/__test__/onDragEnter.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    const myAttachment = new Attachment();
+    const container = myAttachment.render();
+    const drag = container.getElementsByClassName('kuc-attachment-file');
+
+    const dragEvent = createEvent.dragEnter(drag[0]);
+    Object.defineProperty(dragEvent, 'dataTransfer', {
+      value: {}
+    });
+    try {
+      fireEvent(drag[0], dragEvent);
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/onDragEnter.test.ts]: Seem miss some expectation.

File: src/js/Attachment/__test__/onDragEnter.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    const dragEvent = createEvent.dragEnter(drag[0]);
+    Object.defineProperty(dragEvent, 'dataTransfer', {
+      value: {
+        items: [newFile],
+        types: ['123']
+      },
+    });
+    try {
+      fireEvent(drag[0], dragEvent);
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/onDragEnter.test.ts]: Seem miss some expectation.

File: src/js/Attachment/__test__/on.test.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  });
+  afterEach(() => {
+    // @ts-ignore
+    console.error.mockRestore();
+  });
+
+  test('on is called successfully', () => {
+    const myAttachment = new Attachment();
+    myAttachment.on('filesAdd', () => {});
+    myAttachment.on('fileRemove', () => {});
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/on.test.ts]: Miss some expectation.
Need to check this function is called when file is added/removed

File: src/js/Attachment/__test__/hideError.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  test('hideError is called successfully', () => {
+    const ERROR_MESSAGE_TEXT = 'Error.';
+    const myAttachment = new Attachment({isErrorVisible: true, errorMessage: ERROR_MESSAGE_TEXT});
+    const container = myAttachment.render();
+    myAttachment.hideError();
+    const errorDOM = container.getElementsByClassName('kuc-attachment-file-error');
+    expect(errorDOM.length).toEqual(1);
+    expect(errorDOM[0]).toBeInstanceOf(HTMLDivElement);
+    expect((errorDOM[0] as HTMLDivElement).style.display).toEqual('none');
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/hideError.test.ts]: Should change to simple code
expect((errorDOM[0] as HTMLDivElement).style.display).toEqual('block');
==> expect(errorDOM[0]).toHaveStyle('display: block');

File: src/js/Attachment/__test__/onDragEnter.test.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    });
+
+    const dragEvent = createEvent.dragEnter(drag[0]);
+    Object.defineProperty(dragEvent, 'dataTransfer', {
+      value: {
+        items: [newFile]
+      },
+    });
+    try {
+      fireEvent(drag[0], dragEvent);
```
Comment (hokimcuc0202) [inline src/js/Attachment/__test__/onDragEnter.test.ts]: Should expect with DOM when event is fired.

File: src/js/Attachment/__test__/render.test.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      files: [newFile],
+      browseButtonText: BROWSE_BUTTON_TEXT,
+      fileLimitText: FILE_LIMIT_TEXT,
+      dropZoneText: DROPZONE_TEXT,
+      isErrorVisible: true,
+      errorMessage: ERROR_MESSAGE_TEXT
+    });
+    const container = myAttachment.render();
+    expect(container).toBeTruthy();
+    expect(container.className).toContain('kuc-attachment-outer');
```
[the remark this answers is not in the document]
Comment (phongnhat19) [reply src/js/Attachment/__test__/render.test.ts]: Line 41 is used to check if container existed. I think we should have both line 41 and 42 to verify 2 cases.

File: src/js/Attachment/__test__/onDragEnter.test.ts (hunk 4 above)
Comment (phongnhat19) [reply src/js/Attachment/__test__/onDragEnter.test.ts]: This is to verify that `Attachment` component won't crashed (internal function handle that)

File: src/js/Attachment/__test__/onDragEnter.test.ts (hunk 5 above)
Comment (phongnhat19) [reply src/js/Attachment/__test__/onDragEnter.test.ts]: This is to verify that `Attachment` component won't crashed (internal function handle that)

File: src/js/Attachment/__test__/on.test.ts (hunk 6 above)
Comment (phongnhat19) [reply src/js/Attachment/__test__/on.test.ts]: This is to verify that `on` is called successfully to assign event handler. operation of `_onFilesAdd` and `_onFileRemove` are tested in other test suit.

File: src/js/Attachment/__test__/setFiles.test.ts (hunk 2 above)
Comment (phongnhat19) [reply src/js/Attachment/__test__/setFiles.test.ts]: Line 26 and 27 is used verify `name` and `size` property existed before checking the value.

**Pull request merged**